File: film-randomized/src/shared/utils/mediaUtils.js
/**
 * Utility functions for media filtering and selection.
 * Provides helper functions for processing media data from TMDb API.
 */

/**
 * @param {Object} item - Watchlist or discovered row (normalized or API)
 * @returns {'movie'|'tv'|null}
 */
function normalizeItemMediaType(item) {
  let mt = item.media_type ?? item.mediaType;
  if (mt === true) return 'movie';
  if (mt === false) return 'tv';
  if (mt === 'movie' || mt === 'tv') return mt;
  return null;
}

/**
 * @param {boolean|'movie'|'tv'} mediaType - Current generator mode
 * @returns {'movie'|'tv'}
 */
export function resolveMediaTypeString(mediaType) {
  return mediaType === true || mediaType === 'movie' ? 'movie' : 'tv';
}

/**
 * Whether this tmdb id is excluded for the current bucket (same media type only).
 *
 * @param {number} tmdbId - TMDb id for current list (movie or tv)
 * @param {'movie'|'tv'} currentMediaType
 * @param {Array<Object>} excludedItems - Watchlist and/or discovered rows
 * @returns {boolean}
 */
export function isExcludedForCurrentType(tmdbId, currentMediaType, excludedItems) {
  if (!excludedItems || excludedItems.length === 0) return false;
  return excludedItems.some(function (item) {
    const id = item.tmdb_id ?? item.tmdbId;
    const mt = normalizeItemMediaType(item);
    return id === tmdbId && mt === currentMediaType;
  });
}

/**
 * Filters out media in the exclusion list (watchlist + discovered) or without a valid description.
 *
 * @param {Array<Object>} results - Array of media items from API
 * @param {Array<Object>} excludedItems - Items to skip (watchlist + discovered)
 * @param {boolean|'movie'|'tv'} currentMediaType - Active generator mode
 * @returns {Array<Object>} Filtered array of valid media items
 */
export function filterValidMedia(results, excludedItems = [], currentMediaType) {
  const mt = resolveMediaTypeString(currentMediaType);
  return results.filter(function (media) {
    const excluded = isExcludedForCurrentType(media.id, mt, excludedItems);

    return (
      !excluded &&
      media.overview &&
      media.overview.trim() !== '' &&
      media.overview !== 'Nessuna descrizione disponibile in italiano.'
    );
  });
}

/**
 * Checks if a media item has a valid description in Italian.
 * 
 * @param {Object} media - Media object to check
 * @param {string} media.overview - The media overview/description
 * @returns {boolean} True if media has a valid description
 */
export function hasValidDescription(media) {
    return media.overview &&
        media.overview.trim() !== '' &&
        media.overview !== 'Nessuna descrizione disponibile in italiano.';
}

/**
 * Generates a random page number within the total pages.
 * 
 * @param {number} totalPages - Total number of pages available
 * @returns {number} Random page number between 1 and totalPages
 */
export function getRandomPage(totalPages) {
    return Math.floor(Math.random() * totalPages) + 1;
}

/**
 * Selects a random media item from an array.
 * 
 * @param {Array<Object>} mediaArray - Array of media items
 * @returns {Object} Randomly selected media item
 */
export function getRandomMedia(mediaArray) {
    const randomIndex = Math.floor(Math.random() * mediaArray.length);
    return mediaArray[randomIndex];
}

File: film_randomized_mobile/lib/utils/media_utils.dart
import 'dart:math';
import '../models/media_type.dart';

/// Coerces a media type value to canonical string form.
String resolveMediaTypeString(MediaType mediaType) {
  return mediaType == MediaType.movie ? 'movie' : 'tv';
}

/// Whether this tmdb id is excluded for the current bucket (same media type only).
bool isExcludedForCurrentType(
    int tmdbId, String currentMediaType, List<dynamic> excludedItems) {
  if (excludedItems.isEmpty) return false;
  return excludedItems.any((item) {
    final id = item['tmdb_id'] ?? item['tmdbId'];
    final mt = getMediaTypeFromItem(item);
    return id == tmdbId && mt == currentMediaType;
  });
}

/// Checks if a media item has a valid description.
bool hasValidDescription(Map<String, dynamic> media) {
  final overview = media['overview'] as String?;
  if (overview == null || overview.trim().isEmpty) return false;
  if (overview == 'Nessuna descrizione disponibile in italiano.') return false;
  return true;
}

/// Filters out excluded items and items without a valid overview.
List<Map<String, dynamic>> filterValidMedia(
    List<dynamic> results, List<dynamic> excludedItems, MediaType mediaType) {
  final mt = resolveMediaTypeString(mediaType);
  return results.where((media) {
    if (isExcludedForCurrentType(media['id'] as int, mt, excludedItems)) {
      return false;
    }
    return hasValidDescription(media as Map<String, dynamic>);
  }).map((e) => e as Map<String, dynamic>).toList();
}

/// Generates a random page number within the total pages.
int getRandomPage(int totalPages) {
  return Random().nextInt(totalPages) + 1;
}

/// Selects a random media item from an array.
Map<String, dynamic> getRandomMedia(List<Map<String, dynamic>> mediaArray) {
  return mediaArray[Random().nextInt(mediaArray.length)];
}

/// Matches a media item against a collection tab filter.
bool matchesMediaListFilter(Map<String, dynamic> item, String filter) {
  if (filter == 'all') return true;
  final mt = getMediaTypeFromItem(item);
  if (mt == null) return false;
  if (filter == 'movies') return mt == 'movie';
  if (filter == 'tv') return mt == 'tv';
  return true;
}

/// Extracts media type from an item (handles both snake_case and camelCase).
String? getMediaTypeFromItem(Map<String, dynamic> item) {
  final mt = item['media_type'] ?? item['mediaType'];
  if (mt == true) return 'movie';
  if (mt == false) return 'tv';
  if (mt == 'movie' || mt == 'tv') return mt as String;
  return null;
}